pkg/networking/session.go

```diff
@@ -0,0 +1,411 @@
+)
+
+// Session is used to wrap a reliable ordered connection.
+type Session struct {
+	g      *execution.TaskGroup
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	config *Config
+	logger *slog.Logger
+	tp     *timerPool
+
+	conn    io.ReadWriteCloser // conn is the underlying connection
+	bufRead *bufio.Reader      // buffered reader wrapped around the connection
+
+	receiveLock   sync.Mutex    // Guards the receiveBuffer.
+	receiveBuffer *bytes.Buffer // receiveBuffer is used to store the incoming data.
+
+	sendLock sync.Mutex       // Guards the sendCh.
+	sendCh   chan *sendPacket // sendCh is used to send data to the connection.
+
+	establishedLock sync.Mutex // Guards the established field.
+	established     bool       // Indicates that incoming Handshake was successfully accepted.
+	shutdownLock    sync.Mutex // Guards the shutdown field.
+	shutdown        bool       // shutdown is used to safely close the Session.
+}
+
+// NewSession is used to construct a new session.
+func newSession(ctx context.Context, config *Config, conn io.ReadWriteCloser, tp *timerPool) (*Session, error) {
+	if config.protocol == nil {
+		return nil, ErrInvalidConfigurationNoProtocol
+	}
+	if config.handler == nil {
+		return nil, ErrInvalidConfigurationNoHandler
+	}
+	if tp == nil {
+		return nil, ErrEmptyTimerPool
+	}
+	sCtx, cancel := context.WithCancel(ctx)
+	s := &Session{
+		g:       execution.NewTaskGroup(suppressContextCancellationError),
+		ctx:     sCtx,
+		cancel:  cancel,
+		config:  config,
+		tp:      tp,
+		conn:    conn,
+		bufRead: bufio.NewReader(conn),
+		sendCh:  make(chan *sendPacket, 1), // TODO: Make the size of send channel configurable.
+	}
+
+	attributes := []any{
+		slog.String("namespace", Namespace),
+		slog.String("remote", s.RemoteAddr().String()),
+	}
+	attributes = append(attributes, config.attributes...)
+
+	if config.logger == nil {
+		s.logger = slog.Default().With(attributes...)
+	} else {
+		s.logger = config.logger.With(attributes...)
+	}
+
+	s.g.Run(s.receiveLoop)
+	s.g.Run(s.sendLoop)
+	if s.config.keepAlive {
+		s.g.Run(s.keepaliveLoop)
+	}
+
+	return s, nil
+}
+
+func (s *Session) String() string {
+	return fmt.Sprintf("Session{local=%s,remote=%s}", s.LocalAddr(), s.RemoteAddr())
+}
+
+// LocalAddr returns the local network address.
+func (s *Session) LocalAddr() net.Addr {
+	if a, ok := s.conn.(addressable); ok {
+		return a.LocalAddr()
+	}
+	return &sessionAddress{addr: "local"}
+}
+
+// RemoteAddr returns the remote network address.
+func (s *Session) RemoteAddr() net.Addr {
+	if a, ok := s.conn.(addressable); ok {
+		return a.RemoteAddr()
+	}
+	return &sessionAddress{addr: "remote"}
+}
+
+// Close is used to close the session. It is safe to call Close multiple times from different goroutines,
+// subsequent calls do nothing.
+func (s *Session) Close() error {
+	s.shutdownLock.Lock()
+	defer s.shutdownLock.Unlock()
+
+	if s.shutdown {
+		return nil // Fast path - session already closed.
+	}
+	s.shutdown = true
```

Comment:
Ok, I'll update to `sync.Once`, but I don't think it's a good idea to return error if you called `Close` intentionally. The error should be returned only in case of counterparty closed connection.

Reply:
Okay, guess I understand your point of view